### PR TITLE
feat: Add new new push notification interface `ParseNotification` for managing push notifications

### PR DESCRIPTION
**Pull request by mbfakourii**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

## Issue

Closes: #936

## Approach

In the new method, `ParseNotification` should be given to `ParsePush`

```dart
ParsePush.instance.initialize(
  firebaseMessaging,
  parseNotification: ParseNotification(onShowNotification: (value) {
    // show notification
  }),
);
```

## Tasks

- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)

### CODE REVIEW

**Comment by mtrezza** [comment]
@cool2apps Could you test this PR out?

**Comment by mbfakourii** [comment]
> @cool2apps Could you test this PR out, if it works for you?

I think it is ready for merge, it is no longer dependent on another library

**Review by mtrezza — CHANGES_REQUESTED** [review]
Could you please change the PR title? It currently seems to describe the change, but it should describe the issue that is being fixed.

**Comment by mbfakourii** [comment]
> Could you please change the PR title? It currently seems to describe the change, but it should describe the issue that is being fixed.

I changed. I had no better idea for the title

**Comment by mtrezza** [comment]
The PR title must describe the issue, the solution of how it was fixed is an option addition. It's still not clear what this PR fixes. Is it not "properly" displaying notifications - in that case could we specify what that means? Or is it something else?

**Comment by mbfakourii** [comment]
> The PR title must describe the issue, the solution of how it was fixed is an option addition. It's still not clear what this PR fixes. If it'S not "properly" displaying notifications, could we specify what that means? Or is it something else?

I changed it, what do you think?

**Comment by mtrezza** [comment]
Again, this describes the change, not the issue that this PR fixes.

**Comment by mbfakourii** [comment]
> Again, this describes the change, not the issue that this PR fixes.
> 
> * Is this PR fixing a bug, then what does it fix?
> * Is the PR adding a new or changing an existing feature and _not_ fixing a bug, then what is that feature?

In this PR, we solved the problem of using the notification library overhead by removing it and replacing it with an interface.
I think this PR solves a bug

**Comment by mbfakourii** [comment]
> If it was a bug we should be able to describe what's broken or not working, but that seems not possible. If it removes a overhead then maybe it's a performance improvement?
> 
> How about:
> 
> > perf: Remove push notification library overhead by replacing it with a push notification interface
> 
> Just to confirm again, this PR does not require any changes in a developer's app code, right?

Changes are required on the developer side
I think this is a Breaking changes

**Comment by mtrezza** [comment]
We do not distinguish between "small" and "large" breaking changes. Any breaking change usually requires a major version increment, the exception being security fixes which are evaluated on a per-case basis.

> but since ParsePush is a new feature

So then this is would be a `feat` PR and a breaking change. A `feat` is higher in priority than a `perf` change, but we can add the performance implication in the changelog entry.

**Comment by mbfakourii** [comment]
> We do not distinguish between "small" and "large" [breaking changes.](https://github.com/parse-community/parse-server/blob/alpha/CONTRIBUTING.md#breaking-change-1) Any breaking change usually requires a major version increment, the exception being security fixes which are evaluated on a per-case basis.
> 
> > but since `ParsePush` is a new feature
> 
> So then this is would be a `feat` PR that contains a breaking change. A `feat` is higher in priority than a `perf` change, but we can add the performance implication in the changelog entry.
> 
> Merging a breaking change as non-breaking in flutter is especially bad due to the de-facto standard of using range operators for dependency versions. That means they will auto-update and potentially break apps. We need to be extra careful because of that.

I got it, so it is a breaking change and `feat` added in title

**Comment by mtrezza** [comment]
I think the PR title could be improved; with "user interface" you are probably referring to the programming interface, not a UI.

How about this change log entry:

> BREAKING CHANGES:
> 
> - The push notification library `flutter_local_notifications` is replaced with the new push notification interface `ParseNotification` ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
> 
> Features
> - Add new interface `ParseNotification` for managing push notifications ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))

Feel free to add any info to the "breaking change" line that may help developers to cope with the braking change.

**Comment by mbfakourii** [comment]
> I think the PR title could be improved; with "user interface" you are probably referring to the programming interface, not a UI.
> 
> How about this change log entry:
> 
> > BREAKING CHANGES
> > 
> > * The push notification library `flutter_local_notifications` is replaced with the new push notification interface `ParseNotification` ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
> > 
> > Features
> > 
> > * Add new new push notification interface `ParseNotification` for managing push notifications ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
> 
> Feel free to add any info to the "breaking change" line that may help developers to cope with the braking change.

I changed it, what do you think?

**Comment by mtrezza** [comment]
We would need a changelog entry as well

File: packages/flutter/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.1.2...flutter-6.0.0) (2023-08-02)
```
Comment (mtrezza) [inline packages/flutter/CHANGELOG.md:1]:
```suggestion
## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.1.2...flutter-6.0.0) (2023-08-06)
```

File: packages/flutter/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.1.2...flutter-6.0.0) (2023-08-06)
+
+### BREAKING CHANGES
+
+* Add new new push notification interface `ParseNotification` for managing push notifications. ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
+
```
Comment (mtrezza) [inline packages/flutter/CHANGELOG.md:6]:
```suggestion

### Features

* Add new new push notification interface ParseNotification for managing push notifications ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))

```

File: packages/flutter/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,13 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.1.2...flutter-6.0.0) (2023-08-06)
+
+### BREAKING CHANGES
+
+* Add new new push notification interface `ParseNotification` for managing push notifications. ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
```
Comment (mtrezza) [inline packages/flutter/CHANGELOG.md:5]:
```suggestion
* The push notification library flutter_local_notifications is replaced with the new push notification interface ParseNotification ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
```

File: packages/flutter/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,13 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/flutter-5.1.2...flutter-6.0.0) (2023-08-06)
+
+### BREAKING CHANGES
+
+* The push notification library flutter_local_notifications is replaced with the new push notification interface ParseNotification ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
+
+### Features
+
+* Add new new push notification interface ParseNotification for managing push notifications ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
```
Comment (mtrezza) [inline packages/flutter/CHANGELOG.md:9]:
```suggestion
* The push notification library flutter_local_notifications is replaced with the new push notification interface `ParseNotification` ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))

### Features

* Add new new push notification interface `ParseNotification` for managing push notifications ([#949](https://github.com/parse-community/Parse-SDK-Flutter/pull/949))
```

**Comment by mtrezza** [comment]
@mbfakourii I've edited the changelog; to better explain the format, when there's a breaking change, there is the `BREAKING CHANGES` section that can be more verbose and explain what is breaking, how the developer can transition and what to be careful about. Then there is still the `Features` section that has the shorter entry for the feature change. In other words, a PR always requires a base entry like `Bug Fixes` or `Feature` and if it's breaking it requires an additional entry in `BREAKING CHANGES`.

**Review by mtrezza — APPROVED** [review]
Looks good

**Pull request merged**